### ISSUE TITLE
Placeholder existence check

### ISSUE THREAD

**Issue by robwierzbowski** [sass/sass#901]
Coming from https://github.com/nex3/sass/pull/821, can placeholder existance checks be added? Checking existence, especially within the current (`@media`) scope, would be useful.

**Comment by nex3** [comment]
Checking the existence of selectors is very tricky. With things like variables and mixins, when you check the existence you're asking the question, "if I use this, will it break?". But it's perfectly valid for you to extend a placeholder and then define it later on. It's impossible to detect in general whether a placeholder will be defined later on in the file, which means that a `placeholder-defined` function would work differently from all the other definition-checking functions in a reasonably subtle way.

Given these constraints, and the fact that you can just do `@extend %placeholder !optional` to avoid getting undefined-selector warnings, I don't think this is something we're likely to support.

**Comment by chriseppstein** [comment]
I share @nex3's perspective here. Scripting happens before all the selectors exist and so a script check for this seems like an inversion of control. We will have other ways of managing the ability to use `@extend` in `@media` in 3.3.

**Comment by robwierzbowski** [comment]
That's reasonable, no worries. My use was centered around `@media`, building a mixin wrapping mixin that says "placeholder exists? extend placeholder. !placeholder exists? create new placeholder with mixin content". @chriseppstein, can you point me to the 3.3 issue on extend/mqs?

**Comment by lukewatts** [comment]
Apologies if this is a stupid question but why is this closed? I don't see a placeholder-exists function, only variable | function | mixin-exists.

Checking for placeholders would be useful when you mixin uses placeholder extends to keep DRY css and abstract functionality away from the mixin. However, the mixin would fail without them; so checking is required to give an appropriate error or a possible on-the-fly fix.

Does one of the other checks also check for placeholders? Like mixin-exists?

**Comment by chriseppstein** [comment]
Not all Sass concepts are created equally. Placeholders are a kind of selector -- they are not part of SassScript. If we add this, we would also be able to add `classname-exists()`, `id-exists()`, etc. 

@robwierzbowski @lukewatts The right way to do this is to track whether you have created the placeholder in question by storing that state a global variable.

**Comment by lukewatts** [comment]
@chriseppstein Hmmm. Not ideal but I do understand that placeholders are a different construct than functions and mixins, and it's probably not a priority. I often use mixins with extends and it just bugs me I can't existence check for them and offer a graceful alternative if they're not there. Thanks for clarifying though.

**Comment by esr360** [comment]
For what it's worth, he's my small requirement for a 'placeholder-exists' function.

I have a mixin where where the passed content is stored in a randomly generated placeholder, so it can be extended where needed (reference: https://github.com/esr360/Synergy/blob/master/src/scss/mixins/_option.scss#L33). It is technically not impossible for the generated placeholder name to not be unique (as unlikely as it is). A simple if placeholder-exists function would remove the risk of this generated placeholder not being unique.

And whilst we're at it:

> If we add this, we would also be able to add classname-exists(), id-exists(), etc.

Well hot damn, and this is bad because?

**Comment by brookjordan** [comment]
I'd love to see something like this too.
Even if it's named more explicitly, (something like `placeholder-already-exists`), to make it clear that it may exist later.

**Comment by chriseppstein** [comment]
@brookjordan It's better to discuss this issue with class names than with placeholders (because the same rules will apply to both). Then you will probably see this is still very complicated:

* What does it mean for a selector to exist?
   * Does it mean there are styles assigned to an element matching some selector?
   * What if a selector is mentioned in compound selector (E.g. does `.bar` exist because a selector `.foo .bar #baz` was defined?)
* Is it expected by most users that this function would change to return false if the selector was moved to a different css file that is also expected to be linked to the same page?
* Does this function return `true` if the selector's ruleset is empty causing it to be pruned from the output?
* What does it mean for an argument that is a compound or a complex selector?
* Do people expect this to take into account the result of an `@extend` operation that is encountered previously to the function call?

All of these questions (and more) need to be answered before we can add such a feature. If you want to propose a general purpose selector check, please use the [RFC process](https://github.com/sass/language#this-repository).

**Comment by robwierzbowski** [comment]
BTW, I no longer have a need for this feature. Removing myself from the discussion 👋🏼

**Comment by brookjordan** [comment]
Appologies, my reply appears much more rushed than I remembered.

When I said 'already', I was assuming SASS is compiled in a set order. If that's the case, the 'already' means 'by this point in the compilation process'.

I guess the same way as a variable may change later, this function may change value if called multiple times as it would be statefully aware.

I, personally, would be happy with an exact matcher.

I actually wasn't aware that `.foo %bar #baz` was valid, but if it is then that could complicate it... and despite an exact matcher being fine for me, I understand that it may not be for others.

I'll see if I can cobble together something that resembles an RFC. Probably something lower level that would allow a function like this to be built.

**Comment by nex3** [comment]
I'm pretty firmly against this feature, so I'm not sure it's worth the effort to create an RFC. Even if you can answer all those questions in a way that makes sense *to you*, the fact that they can all be raised indicates that the behavior of the function is likely to be unintuitive to many users no matter which answers are chosen. The fact is that the notion of whether styles are defined for a given selector just doesn't match up well with Sass's procedural compilation mode, which makes functions like this inherently misleading.

**Comment by brookjordan** [comment]
What if I can answer them in a way that makes sense with the way SASS works in other places?
I'm trying to work out how I can put it more on par with how variables work. That is: they are the value that they are at the point that they are used but can print a different value later.
As such, one of my thoughts was some type of tagged global variable that would keep track. I'm not sure how to deal with that as of yet, but I don't feel that it's undoable in a SASS-like way.

Could I ask why you're so firmly against the feature? Which parts in particular you dislike?
I'm guessing that my message came across in the wrong way because I mentioned myself so much. This was me trying to clarify that I understand that my needs are not the same as the needs of everyone. That's why my intention was to design something lower-level which should fit the current paradigm of the system.

To put this in perspective, I'm building quite a complicated project: a game-engine using only CSS and HTML. It's coming along nicely, and I'm finding work-arounds all over the place. But having detection for these place-holders would mean that a user of the engine could override defaults rather than importing the defaults they need.
Perhaps I could walk through with you my problem at some point and you could possibly come up with some better solutions?
It's very possible I've missed something I could use instead.

**Comment by nex3** [comment]
The problem with trying to do the same thing as we do with variables is that style rules fundamentally don't work the same way Sass variables do. They're declarative, they cascade, and they apply selectively based on the structure of the document which isn't known at compile-time. All of this makes the tools we use for dealing with variables inapplicable to selectors. Put another way, the question "has Sass compiled a style rule that contains this simple selector yet" doesn't map to any semantic insight about the way the document is being styled.

> To put this in perspective, I'm building quite a complicated project: a game-engine using only CSS and HTML. It's coming along nicely, and I'm finding work-arounds all over the place. But having detection for these place-holders would mean that a user of the engine could override defaults rather than importing the defaults they need.
> Perhaps I could walk through with you my problem at some point and you could possibly come up with some better solutions?
> It's very possible I've missed something I could use instead.

Style rules have a built-in mechanism for overriding defaults: the cascade. I recommend you use that, possibly wrapped with mixins if you want to control the specifics of what people can override.
